File: packages/gbnf-experiment/data/20260909T144331Z_5b46f630/ported_implementation/src/grammarParser/buildRuleStack.ts
import {
  RuleChar,
  RuleCharExclude,
  RuleEnd,
  type Range,
  type UnresolvedRule,
} from '../grammarGraph/grammarGraphTypes.js';
import { RuleRef } from '../grammarGraph/ruleRef.js';
import { isRange, isRuleEnd } from '../grammarGraph/typeGuards.js';
import {
  isRuleDefAlt,
  isRuleDefChar,
  isRuleDefCharAlt,
  isRuleDefCharNot,
  isRuleDefCharRngUpper,
  isRuleDefEnd,
  isRuleDefRef,
  type InternalRuleDef,
  type InternalRuleDefChar,
  type InternalRuleDefCharNot,
} from '../rulesBuilder/rulesBuilderTypes.js';

export const makeCharRule = (
  ruleDef: InternalRuleDefChar | InternalRuleDefCharNot,
): RuleChar | RuleCharExclude => {
  const value = ruleDef.value as (number | Range)[];
  if (isRuleDefCharNot(ruleDef)) {
    return new RuleCharExclude(value);
  }
  if (isRuleDefChar(ruleDef)) {
    return new RuleChar(value);
  }

  throw new Error(`Unsupported rule type for make_char_rule: ${ruleDef}`);
};

export const buildRuleStack = (
  linearRules: InternalRuleDef[],
): UnresolvedRule[][] => {
  let paths: UnresolvedRule[] = [];
  const stack: UnresolvedRule[][] = [];
  let idx = 0;

  while (idx < linearRules.length) {
    const ruleDef = linearRules[idx];
    if (isRuleDefChar(ruleDef) || isRuleDefCharNot(ruleDef)) {
      // this could be a single char, or a range, or a sequence of alts; we don't know until we
      // step through it.
      const charRule = makeCharRule(ruleDef);
      idx += 1;
      let rule: InternalRuleDef | null = idx < linearRules.length ? linearRules[idx] : null;
      while (
        idx < linearRules.length &&
        (isRuleDefCharRngUpper(rule) || isRuleDefCharAlt(rule))
      ) {
        if (isRuleDefCharRngUpper(rule)) {
          // previous rule value should be a number
          const prevValue = charRule.value.pop();
          if (isRange(prevValue)) {
            throw new Error(
              `Unexpected range, expected a number but got an array: ${JSON.stringify(prevValue)}`,
            );
          }
          if (prevValue === undefined) {
            throw new Error('Unexpected undefined value');
          }

          charRule.value.push([prevValue, rule.value]);
        }
        if (isRuleDefCharAlt(rule)) {
          charRule.value.push(rule.value);
        }
        idx += 1;
        rule = idx < linearRules.length ? linearRules[idx] : null;
      }
      paths.push(charRule);
    } else {
      if (isRuleDefAlt(ruleDef)) {
        if (paths.length === 0) {
          throw new Error('Encountered alt without anything before it');
        }
        paths.push(new RuleEnd());
        stack.push(paths);
        paths = [];
      } else if (isRuleDefEnd(ruleDef)) {
        paths.push(new RuleEnd());
      } else if (isRuleDefRef(ruleDef)) {
        paths.push(new RuleRef(ruleDef.value));
      } else if (isRuleDefCharAlt(ruleDef)) {
        throw new Error(
          `Encountered char alt, should be handled by above block: ${ruleDef}`,
        );
      } else {
        throw new Error(`Unsupported rule type: ${ruleDef}`);
      }
      idx += 1;
    }
  }

  if (!isRuleEnd(paths[paths.length - 1])) {
    paths.push(new RuleEnd());
  }

  stack.push(paths);
  return stack;
};

File: proposed-projects/round-trip-experiment/reverse/20260909T175143Z_62b61540/20260910T094302Z_8f8be025/ported_implementation/gbnf/grammar_parser/build_rule_stack.py
import json
from typing import Optional, Union

from ..grammar_graph.grammar_graph_types import RuleChar, RuleCharExclude, RuleEnd
from ..grammar_graph.rule_ref import RuleRef
from ..grammar_graph.type_guards import is_range, is_rule_end
from ..rules_builder.rules_builder_types import (
    InternalRuleDef,
    is_rule_def_alt,
    is_rule_def_char,
    is_rule_def_char_alt,
    is_rule_def_char_not,
    is_rule_def_char_rng_upper,
    is_rule_def_end,
    is_rule_def_ref,
)


def _serialize(rule_def: InternalRuleDef) -> str:
    return json.dumps(rule_def, separators=(",", ":"), default=str)


def _at(rules: list[InternalRuleDef], idx: int) -> Optional[InternalRuleDef]:
    if 0 <= idx < len(rules):
        return rules[idx]
    return None


def make_char_rule(rule_def: InternalRuleDef) -> Union[RuleChar, RuleCharExclude]:
    if is_rule_def_char_not(rule_def):
        return RuleCharExclude(rule_def["value"])
    if is_rule_def_char(rule_def):
        return RuleChar(rule_def["value"])

    raise ValueError(
        f"Unsupported rule type for make_char_rule: {_serialize(rule_def)}"
    )


def build_rule_stack(linear_rules: list[InternalRuleDef]) -> list[list]:
    paths: list = []
    stack: list[list] = []
    idx = 0

    while idx < len(linear_rules):
        rule_def = linear_rules[idx]
        if is_rule_def_char(rule_def) or is_rule_def_char_not(rule_def):
            # this could be a single char, or a range, or a sequence of alts;
            # we don't know until we step through it.
            char_rule = make_char_rule(rule_def)
            idx += 1
            rule = _at(linear_rules, idx)
            while idx < len(linear_rules) and (
                is_rule_def_char_rng_upper(rule) or is_rule_def_char_alt(rule)
            ):
                if is_rule_def_char_rng_upper(rule):
                    # previous rule value should be a number
                    prev_value = char_rule.value.pop() if char_rule.value else None
                    if is_range(prev_value):
                        raise ValueError(
                            "Unexpected range, expected a number but got an array: "
                            f"{json.dumps(prev_value, separators=(',', ':'))}"
                        )
                    if prev_value is None:
                        raise ValueError("Unexpected undefined value")

                    char_rule.value.append([prev_value, rule["value"]])
                if is_rule_def_char_alt(rule):
                    char_rule.value.append(rule["value"])
                idx += 1
                rule = _at(linear_rules, idx)
            paths.append(char_rule)
        else:
            if is_rule_def_alt(rule_def):
                if len(paths) == 0:
                    raise ValueError("Encountered alt without anything before it")
                paths.append(RuleEnd())
                stack.append(paths)
                paths = []
            elif is_rule_def_end(rule_def):
                paths.append(RuleEnd())
            elif is_rule_def_ref(rule_def):
                paths.append(RuleRef(rule_def["value"]))
            elif is_rule_def_char_alt(rule_def):
                raise ValueError(
                    "Encountered char alt, should be handled by above block: "
                    f"{_serialize(rule_def)}"
                )
            else:
                raise ValueError(f"Unsupported rule type: {_serialize(rule_def)}")
            idx += 1

    if not is_rule_end(paths[-1] if paths else None):
        paths.append(RuleEnd())

    stack.append(paths)
    return stack
